test/version_solver_test.dart

```diff
@@ -1134,9 +1134,89 @@ void dartSdkConstraint() {
 
     await expectResolves(
         environment: {'_PUB_TEST_SDK_VERSION': '2.0.0'},
-        error: 'Package foo requires SDK version <2.0.0-dev.infinity but the '
+        error: 'Package foo requires SDK version <2.0.0 but the '
             'current SDK is 2.0.0.');
   });
+
+  test("prerelease override requires major sdk versions to match", () async {
+    await d.dir(appPath, [
+      await d.pubspec({
+        'name': 'myapp',
+        'environment': {'sdk': '<2.2.3'}
+      })
+    ]).create();
+
+    await expectResolves(
+        environment: {'_PUB_TEST_SDK_VERSION': '1.2.3-dev.1.0'},
+        output: isNot(contains('PUB_ALLOW_PRERELEASE_SDK')));
+  });
+
+  test("prerelease override requires minor sdk versions to match", () async {
+    await d.dir(appPath, [
+      await d.pubspec({
+        'name': 'myapp',
+        'environment': {'sdk': '<1.3.3'}
+      })
+    ]).create();
+
+    await expectResolves(
+        environment: {'_PUB_TEST_SDK_VERSION': '1.2.3-dev.1.0'},
+        output: isNot(contains('PUB_ALLOW_PRERELEASE_SDK')));
+  });
+
+  test("prerelease override requires patch sdk versions to match", () async {
+    await d.dir(appPath, [
+      await d.pubspec({
+        'name': 'myapp',
+        'environment': {'sdk': '<1.2.4'}
+      })
+    ]).create();
+
+    await expectResolves(
+        environment: {'_PUB_TEST_SDK_VERSION': '1.2.3-dev.1.0'},
+        output: isNot(contains('PUB_ALLOW_PRERELEASE_SDK')));
+  });
+
+  test("prerelease override requires exclusive max", () async {
+    await d.dir(appPath, [
+      await d.pubspec({
+        'name': 'myapp',
+        'environment': {'sdk': '<=1.2.3'}
+      })
+    ]).create();
+
+    await expectResolves(
+        environment: {'_PUB_TEST_SDK_VERSION': '1.2.3-dev.1.0'},
+        output: isNot(contains('PUB_ALLOW_PRERELEASE_SDK')));
+  });
+
+  test("prerelease override requires prerelease sdk", () async {
+    await d.dir(appPath, [
+      await d.pubspec({
+        'name': 'myapp',
+        'environment': {'sdk': '<1.2.3'}
+      })
+    ]).create();
+
+    await expectResolves(
+        environment: {'_PUB_TEST_SDK_VERSION': '1.2.3'},
+        error: 'Package myapp requires SDK version <1.2.3 but the current '
+            'SDK is 1.2.3.');
+  });
+
+  test("prerelease override works generally", () async {
+    await d.dir(appPath, [
+      await d.pubspec({
+        'name': 'myapp',
+        'environment': {'sdk': '<1.2.3'}
+      })
+    ]).create();
+
+    await expectResolves(
+        environment: {'_PUB_TEST_SDK_VERSION': '1.2.3-dev.1.0'},
+        output: allOf(contains('PUB_ALLOW_PRERELEASE_SDK'),
+            contains('<=1.2.3-dev.1.0'), contains('myapp')));
+  });
```

Comment:
It might be worth adding a test for how it works with a "+..." build version too. The SDK doesn't use those now, but just in case we do in the future.

Reply:
done, added one to confirm it doesn't override if there is a dev release constraint explicitly as well since I forgot that